### PR TITLE
US796 Upgrade Manager Folder Refactor

### PR DESCRIPTION
**Pull request by dzadroga**
US796 refactored the Upgrade Manager folder to remove the unnecessary strategy patterns and to simplify the code. 

The audit can be found here: https://docs.google.com/document/d/1OWn_M8yMfUUCEhBGecR1QhP47lW2dATJamBUtd_I0H0/edit?usp=sharing

Tasks:
- Delete IUpgradeStrategy and StandardUpgradeStrategy, remove any references.  (Task 797)
-  Create a static upgrade class that had upgrade decision and execution logic (Task 798)
- Update IUpgradeableTileInfo to make upgrade behaviour explicit (Task 799)
- Create UpgradeRequestEvent that contains PlayerID or TileId (Task 800)
- Create UpgradeResultEvent that has success/failure, failure reason, upgrade cost, new upgrade level (Task 801)
- Make UpgradeManager subscribe to UpgradeRequestEvent, process logic and publish to UpgradeResultEvent (Task 802)
- Add UpgradeFailReason enum to UpgradeDecision to describe why the upgrade failed (Task 803)
- Remove OwnableSpaceTileAdapter and refactor upgrade logic to work directly with PropertySpaceData (Task 804)
- Make sure upgrade requests are only processed if the requesting player owns the property. (Task 805)
- Implement missing rules checks: Player owns the full monopoly before upgrading and house are upgrade evenly across  monopoly group. (Task 806)
- Write Unit Tests for upgrade decision logic (Task 807)

Branch was rebased on dev on 3/28/2026 around 4pm est so no merge conflicts should be present and all tests are passing.

### CODE REVIEW

File: Assets/Scripts/Managers/Upgrade/IUpgradableTileInfo.cs [hunk 1]
```diff
@@ -1,12 +1,20 @@
 using Assets.Scripts.Managers.Rent;
-using UnityEngine;
 
-public interface IUpgradableTileInfo : ITileRentInfo
+public interface IUpgradableTileInfo
```
Comment (hdathert) [inline Assets/Scripts/Managers/Upgrade/IUpgradableTileInfo.cs]: Why do we still have this adapter interface when all of the data already existed in PropertySpaceData? It seems weird to me that we're basically implementing a series of interface getters on property space data for whats already there- not adding functionality, but simply renaming attributes that already existed. It just seems like we're adding complication to PropertySpaceData to justify this interfaces existence from what I'm seeing.

File: Assets/Scripts/Managers/Upgrade/UpgradeManager.cs [hunk 2]
```diff
@@ -13,40 +33,89 @@ private void Awake()
...
+        return null;
+    }
+
+    private IUpgradableTileInfo ResolveTile(int tileId)
+    {
+        //TODO hook into tile 
+        return null;
+    }
+
+    private IUpgradableTileInfo[] GetMonopolyGroup(
```
Comment (hdathert) [inline Assets/Scripts/Managers/Upgrade/UpgradeManager.cs]: This seems to duplicate the functionality of `Player.GetNumberOfOwnedPropertiesByColor()` or `Player.GetValidUpgradableProperties()` but requires us to keep a completely separate list of properties in memory- which means there might end up being multiple sources of truth, depending on how we populate that array.

File: Assets/Scripts/Events/EventChannelTypes/UpgradeRequestEventChannel.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+using UnityEngine;
+
+public class UpgradeRequestEventChannel : ScriptableObject
```
Comment (hdathert) [inline Assets/Scripts/Events/EventChannelTypes/UpgradeRequestEventChannel.cs]: This is not using the same event channel pattern we've established already- reference `BooleanEventChannel.cs` for example on usage.

File: Assets/Scripts/Events/EventChannelTypes/UpgradeResultEventChannel.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+using UnityEngine;
+
+public class UpgradeResultEventChannel : ScriptableObject
```
Comment (hdathert) [inline Assets/Scripts/Events/EventChannelTypes/UpgradeResultEventChannel.cs]: This is not using the same event channel pattern we've established already- reference `BooleanEventChannel.cs` for example on usage.

File: Assets/Scripts/Events/EventDataStructures/UpgradeResultEvent.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        int newUpgradeLevel,
+        int playerId,
+        int tileId)
+    {
+        Success = success;
+        FailReason = failReason;
+        UpgradeCost = upgradeCost;
+        NewUpgradeLevel = newUpgradeLevel;
+        PlayerId = playerId;
+        TileId = tileId;
```
Comment (hdathert) [inline Assets/Scripts/Events/EventDataStructures/UpgradeResultEvent.cs]: What is the `tileID` for? We don't use IDs for spaces anywhere else in the system... we should probably just be sending a reference to the property space itself.

File: Assets/Scripts/Managers/Upgrade/UpgradeManager.cs [hunk 6]
```diff
@@ -13,40 +33,89 @@ private void Awake()
...
-        if (strategy == null)
-        {
-            strategy = new StandardUpgradeStrategy();
-        }
-        // TODO: Any channels to be subscribe to add it here.
+        //TODO: Any channels to be subscribe to add it here.
     }
 
-    public bool TryHandleUpgrade(Player owner, IUpgradableTileInfo tile, out UpgradeDecision decision)
+    public bool TryHandleUpgrade(Player owner, IUpgradableTileInfo tile, OwnableSpaceData[] allSpaces, out UpgradeDecision decision)
```
Comment (hdathert) [inline Assets/Scripts/Managers/Upgrade/UpgradeManager.cs]: What is the use-case difference between `TryHandleUpgrade()` vs `OnUpgradeRequest()`? They appear to be doing the same things, but with different signatures.

Is the intended flow to call the manager directly, or for the manager to be called via events?

**Review by hdathert — CHANGES_REQUESTED** [review]
Just have a few questions/concerns on this one- comments below.

**Review by hdathert — APPROVED** [review]
# Code Review Checklist

**Completed for**: US #796 and Tasks #797 to #807  
**Code Review Completed By**: @hdathert 

Use this checklist to review new code changes before merging.

---

## Implementation
- [x] Does this code change accomplish what it is supposed to do?
- [x] Can the solution be simplified?
- [x] This change does **not** add unwanted compile-time or run-time dependencies.
- [x] The code is at the right abstraction level.
- [x] The code is modular enough.
- [x] The solution is adequate for maintainability, readability, performance, and security.
- [x] Similar functionality does **not** already exist in the codebase (or reuse is not appropriate).
- [x] No additional best practices or design patterns would substantially improve this code.

---

## Logic Errors and Bugs
- [x] No use cases where the code behaves incorrectly.
- [x] No inputs or external events that can break the code.
- [x] Code agrees with specifications.
- [x] No off-by-one errors.
- [x] No magic numbers (or they are well-named constants).
- [x] Proper use of global/large-scope variables (minimized and justified).

---

## Error Handling and Logging
- [x] Error handling is implemented correctly.
- [x] Error messages are user-friendly (where applicable).
- [x] Sufficient, meaningful log events exist for debugging (no sensitive data in logs).

---

## Dependencies
- [x] Documentation/configuration/README updates were made as needed.
- [x] No negative impact on other parts of the system or backward compatibility.

---

## Security and Data Privacy
- [x] No new security vulnerabilities introduced.
- [x] All user input is validated, sanitized, and escaped (XSS/SQL injection prevented).
- [x] Data from external APIs/libraries is handled safely and checked as needed.

---

## Performance
- [x] No significant performance regressions.
- [x] No obvious opportunities for significant performance gains left unaddressed.

---

## Ethics and Morality
- [x] Appropriate measures exist to prevent or report harassment/abuse (if applicable).

---

## Testing and Testability
- [x] Code is testable.
- [x] Automated tests have been added or updated.
- [x] Existing tests sufficiently cover the change (unit/integration/system).
- [x] Edge cases and boundary inputs are tested.

---

## Readability
- [x] Code is easy to understand.
- [x] Methods/functions are appropriately small and focused.
- [x] Names (functions/classes/variables) are clear and descriptive.
- [x] Code is in the correct file/folder/package.
- [x] Control flow is intuitive; data flow is clear.
- [x] Comments are clear, relevant, and up to date.
- [x] No commented-out or redundant code remains.
- [x] No overly dense one-liners; no overly long methods.

---

**Use this before submitting a pull request.**  
It helps ensure code quality, maintainability, and compliance with best practices.

*Based on awesome code review guides and MIT’s code review recommendations.*
